doc/man/nsdb-resolve-fsn.8
.\"@(#)nsdb-resolve-fsn.8"
.\"
.\" @file doc/man/nsdb-resolve-fsn.8
.\" @brief man page for nsdb-resolve-fsn client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH NSDB-RESOLVE-FSN 8 "@publication-date@"
.SH NAME
nsdb-resolve-fsn \- resolve a fileset name (FSN) record on an NSDB
.SH SYNOPSIS
.B nsdb-resolve-fsn
.RB [ \-?d ]
.RB [ \-e
.IR nce ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.I fsn-uuid
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
The bulk of FedFS junction information in a FedFS domain is stored
on one or more LDAP servers.
These servers are known as
.IR "namespace databases" ,
or NSDBs, for short.
.P
FedFS-enabled file servers and clients access the information stored
on NSDBs via standard LDAP queries.
FedFS-enabled file servers use these queries to resolve FedFS junctions.
FedFS administrators use them to manage information
about file sets contained in a FedFS domain name space.
.SH DESCRIPTION
The
.BR nsdb-resolve-fsn (8)
command is part of a collection of low-level single-use programs that are
intended for testing the NSDB protocol or for use in scripts.
This command resolves a FedFS
.I fileset name
(FSN) record on an NSDB into a list of fileset locations.
.P
A fileset name, or FSN, uniquely identifies a fileset in FedFS.
An FSN consists of a UUID and the hostname and port of an NSDB.
This pair is intended to be unique across all of FedFS.
The meaning of these items is described in more detail in
.BR fedfs (7).
.P
A FedFS junction contains an FSN.
There can be multiple junctions that contain a particular FSN.
There is exactly one FSN record stored on an NSDB that corresponds to this FSN.
The FSN record can have zero or more FSL records as children.
Replicas of these records can exist on more than one LDAP server.
.P
The
.BR nsdb-resolve-fsn (8)
command looks up an FSN record on the named NSDB
and returns the set of FSL records that are its children.
This is the same operation that FedFS-enabled file servers perform
when resolving the FSN contained in a FedFS junction.
.P
This command has one positional parameter which specifies
the UUID of the FSN record to resolve.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR nsdb-resolve-fsn (8)
version information and a usage message on
.IR stderr .
.IP "\fB-e, \-\-nce=\fINSDB-container-entry-distinguished-name\fP"
Specifies the distinguished name of the NSDB Container Entry
under which this FSN record resides.
If the
.B \-\-nce
option is not specified,
the value of the FEDFS_NSDB_NCE environment variable is consulted.
If this variable is not set,
then the NSDB connection parameter database is searched for this DN.
If none of these is specified, the
.BR nsdb-resolve-fsn (8)
command searches the NSDB's naming contexts to discover its NCEs.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-host-name\fP"
Specifies the NSDB hostname portion of the FSN to resolve.
If the
.B \-\-nsdbname
option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B \-\-nsdbname
option is not specified, the
.BR nsdb-resolve-fsn (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the NSDB IP port portion of the FSN to resolve.
If the
.B \-\-nsdbport
option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.SH EXIT CODES
The NSDB returns a value that reflects the success of the requested operation.
.TP
.B FEDFS_OK
The LDAP query succeeded.
.TP
.B FEDFS_ERR_ACCESS
The anonymous entity does not have permission to perform the requested operation.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The
.BR nsdb-resolve-fsn (8)
command was unable to find a route to the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The
.BR nsdb-resolve-fsn (8)
command determined that the specified NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The
.BR nsdb-resolve-fsn (8)
command was unable to establish a connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The
.BR nsdb-resolve-fsn (8)
command was unable to authenticate
and establish a secure connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
A non-specific LDAP error occurred on the connection between the
.BR nsdb-resolve-fsn (8)
command and specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
An LDAP error occurred on the connection between the
.BR nsdb-resolve-fsn (8)
command and specified NSDB.
The specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_NONCE
The
.BR nsdb-resolve-fsn (8)
command was unable to locate the NCE on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSN
The
.BR nsdb-resolve-fsn (8)
command was unable to locate the specified FSN on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSL
The
.BR nsdb-resolve-fsn (8)
command was unable to locate any FSLs for the specified FSN
on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The
.BR nsdb-resolve-fsn (8)
command received a malformed response from the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the specified NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The local NSDB connection parameter database
does not have any connection parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The
.BR nsdb-resolve-fsn (8)
command received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
The
.BR nsdb-resolve-fsn (8)
command received an LDAP referral that it was unable to follow.
A specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The
.BR nsdb-resolve-fsn (8)
command received an LDAP referral that it chose not to follow,
either because the local implementation does not support
following LDAP referrals or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The
.BR nsdb-resolve-fsn (8)
command received an LDAP referral that it chose not to follow
because the local NSDB connection parameter database had no
connection parameters for the NSDB targeted by the LDAP referral.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that you have created a new FSN that looks like:
.RS
.sp
	FSN UUID: 8e246ddc-7b46-11e0-8252-000c297fd679
.br
	NSDB: nsdb.example.net:389
.sp
.RE
Further suppose the NSDB
.I nsdb.example.net:389
has an NSDB Container Entry whose distinguished name is
.IR o=fedfs ,
and that the FSN has a single FSL child record.
To resolve the FSN, you might use:
.RS
.sp
$ nsdb-resolve-fsn -e o=fedfs \\
.br
	-l nsdb.example.net \\
.br
	8e246ddc-7b46-11e0-8252-000c297fd679
.sp
For FSN UUID 8e246ddc-7b46-11e0-8252-000c297fd679
.br
    FSN TTL  600
.sp
------------------------------------------------------
.br
dn: fedfsFslUuid=323c5068-7c11-11e0-8d38-000c297fd679,
.br
    fedfsFsnUuid=8e246ddc-7b46-11e0-8252-000c297fd679,o=fedfs
.sp
 FSN UUID:		8e246ddc-7b46-11e0-8252-000c297fd679
.br
 FSL UUID:		323c5068-7c11-11e0-8d38-000c297fd679
.br
 NFS fls_server:		fileserver.example.net
.sp
 NFS fli_rootpath:		/path
.br
 NFS fls_currency:		-1
.sp
.RE
and so on.
.SH SECURITY
The NSDB protocol draft standard requires that FedFS FSN and FSL
records are readable by everyone.
The
.BR nsdb-resolve-fsn (8)
command uses anonymous binding to perform LDAP queries.
.P
The target LDAP server must be registered in the local NSDB connection
parameter database.
The connection security mode listed
in the NSDB connection parameter database
for the target LDAP server is used during this operation.
See
.BR nsdbparams (8)
for details on how to register an NSDB
in the local NSDB connection parameter database.
.SH "SEE ALSO"
.BR fedfs (7),
.BR nsdb-list (8),
.BR nsdbparams (8)
.sp
RFC 5716 for FedFS requirements and overview
.sp
RFC 4510 for an introduction to LDAP
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/nsdbc/nsdb-resolve-fsn.c
/**
 * @file src/nsdbc/nsdb-resolve-fsn.c
 * @brief Resolve an FSN UUID to FSL information using a target NSDB server
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>
#include <netdb.h>
#include <langinfo.h>

#include <uuid/uuid.h>

#include "fedfs.h"
#include "nsdb.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Short form command line options
 */
static const char nsdb_resolve_fsn_opts[] = "?de:l:r:";

/**
 * Long form command line options
 */
static const struct option nsdb_resolve_fsn_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "nce", 1, NULL, 'e', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nsdbport", 1, NULL, 'r', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 */
static void
nsdb_resolve_fsn_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [ -d ] "
			"[ -l nsdbname ] [ -r nsdbport ] [ -e nce ] "
			"fsn-uuid\n\n", progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-e, --nce            DN of NSDB container entry\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

/**
 * Return presentation string for a boolean value
 *
 * @param value a boolean value
 * @return NUL-terminate C string
 */
static const char *
_display_bool(const _Bool value)
{
	return value ? "TRUE" : "FALSE";
}

/**
 * Display nfs_fsl portion of a fedfs_fsl structure
 *
 * @param nfsl pointer to a fedfs_nfs_fsl structure
 */
static void
nsdb_resolve_fsn_display_nfs_fsl(struct fedfs_nfs_fsl *nfsl)
{
	FedFsStatus status;
	char *rootpath;

	status = nsdb_path_array_to_posix(nfsl->fn_nfspath, &rootpath);
	if (status != FEDFS_OK)
		return;

	if (nfsl->fn_fslport == 0)
		printf(" NFS fls_server:\t\t%s\n", nfsl->fn_fslhost);
	else
		printf(" NFS fls_server:\t\t%s:%u\n", nfsl->fn_fslhost,
			nfsl->fn_fslport);

	printf(" NFS fli_rootpath:\t\t%s\n", rootpath);
	free(rootpath);

	printf(" NFS fls_currency:\t\t%d\n", nfsl->fn_currency);
	printf(" NFS FSLI4GF_WRITABLE:\t\t%s\n", _display_bool(nfsl->fn_gen_writable));
	printf(" NFS FSLI4GF_GOING:\t\t%s\n", _display_bool(nfsl->fn_gen_going));
	printf(" NFS FSLI4GF_SPLIT:\t\t%s\n", _display_bool(nfsl->fn_gen_split));
	printf(" NFS FSLI4TF_RDMA:\t\t%s\n", _display_bool(nfsl->fn_trans_rdma));
	printf(" NFS FSLI4BX_CLSIMUL:\t\t%d\n", nfsl->fn_class_simul);
	printf(" NFS FSLI4BX_CLHANDLE:\t\t%d\n", nfsl->fn_class_handle);
	printf(" NFS FSLI4BX_CLFILEID:\t\t%d\n", nfsl->fn_class_fileid);
	printf(" NFS FSLI4BX_CLWRITEVER:\t%d\n", nfsl->fn_class_writever);
	printf(" NFS FSLI4BX_CLCHANGE:\t\t%d\n", nfsl->fn_class_change);
	printf(" NFS FSLI4BX_CLREADDIR:\t\t%d\n", nfsl->fn_class_readdir);
	printf(" NFS FSLI4BX_READRANK:\t\t%d\n", nfsl->fn_readrank);
	printf(" NFS FSLI4BX_READORDER:\t\t%d\n", nfsl->fn_readorder);
	printf(" NFS FSLI4BX_WRITERANK:\t\t%d\n", nfsl->fn_writerank);
	printf(" NFS FSLI4BX_WRITEORDER:\t%d\n", nfsl->fn_writeorder);
	printf(" NFS FSLI4F_VAR_SUB:\t\t%s\n", _display_bool(nfsl->fn_varsub));
	printf(" NFS fli_valid_for:\t\t%d\n", nfsl->fn_validfor);

	printf("\n");
}

/**
 * Display one FSL
 *
 * @param fsl pointer to a fedfs_fsl structure
 */
static void
nsdb_resolve_fsn_display_fsl(struct fedfs_fsl *fsl)
{
	int i;

	printf("------------------------------------------------------\n");
	printf("dn: %s\n\n", fsl->fl_dn);

	printf(" FSN UUID:\t\t%s\n", fsl->fl_fsnuuid);
	printf(" FSL UUID:\t\t%s\n", fsl->fl_fsluuid);

	if (fsl->fl_annotations != NULL) {
		for (i = 0; fsl->fl_annotations[i] != NULL; i++)
			printf(" annotation[%d]: %s\n", i, fsl->fl_annotations[i]);
		printf("\n");
	}

	if (fsl->fl_description != NULL) {
		for (i = 0; fsl->fl_description[i] != NULL; i++)
			printf(" description[%d]: %s\n", i, fsl->fl_description[i]);
		printf("\n");
	}

	if (fsl->fl_type == FEDFS_NFS_FSL)
		nsdb_resolve_fsn_display_nfs_fsl(&fsl->fl_u.fl_nfsfsl);
}

/**
 * Display the returned FSL list
 *
 * @param fsls a list of fedfs_fsl structures
 */
static void
nsdb_resolve_fsn_display_fsls(struct fedfs_fsl *fsls)
{
	for ( ; fsls != NULL; fsls = fsls->fl_next)
		nsdb_resolve_fsn_display_fsl(fsls);
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *nsdbname;
	unsigned short nsdbport;
	struct fedfs_fsl *fsls;
	struct fedfs_fsn *fsn;
	unsigned int ldap_err;
	char *nce, *fsn_uuid;
	FedFsStatus retval;
	int fsn_ttl, arg;
	nsdb_t host;

	(void)umask(S_IRWXO);

	/* Ensure UTF-8 strings can be handled transparently */
	if (setlocale(LC_CTYPE, "") == NULL ||
	    strcmp(nl_langinfo(CODESET), "UTF-8") != 0) {
		fprintf(stderr, "Failed to set locale and langinfo\n");
		exit((int)FEDFS_ERR_INVAL);
	}

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb_env(&nsdbname, &nsdbport, NULL, &nce);

	fsn_uuid = NULL;
	while ((arg = getopt_long(argc, argv, nsdb_resolve_fsn_opts,
			nsdb_resolve_fsn_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'e':
			nce = optarg;
			break;
		case 'l':
			nsdbname = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdbport)) {
				fprintf(stderr, "Bad port number: %s\n",
					optarg);
				nsdb_resolve_fsn_usage(progname);
			}
			break;
		default:
			fprintf(stderr, "Invalid command line "
				"argument: %c\n", (char)arg);
		case '?':
			nsdb_resolve_fsn_usage(progname);
		}
	}
	if (argc == optind + 1) {
		uuid_t uu;
		fsn_uuid = argv[optind];
		if (uuid_parse(fsn_uuid, uu) == -1) {
			fprintf(stderr, "Invalid FSN UUID was specified\n");
			nsdb_resolve_fsn_usage(progname);
		}
	} else if (argc >  optind + 1) {
		fprintf(stderr, "Unrecognized positional parameters\n");
		nsdb_resolve_fsn_usage(progname);
	} else {
		fprintf(stderr, "No FSN UUID was specified\n");
		nsdb_resolve_fsn_usage(progname);
	}
	if (nsdbname == NULL) {
		fprintf(stderr, "No NSDB hostname was specified\n");
		nsdb_resolve_fsn_usage(progname);
	}

	retval = nsdb_lookup_nsdb(nsdbname, nsdbport, &host, NULL);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_PARAMS:
		fprintf(stderr, "No connection parameters for NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out;
	default:
		fprintf(stderr, "Failed to look up NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out;
	}

	retval = nsdb_open_nsdb(host, NULL, NULL, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_CONN:
		fprintf(stderr, "Failed to connect to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_AUTH:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u: %s\n",
			nsdbname, nsdbport, ldap_err2string(ldap_err));
		goto out_free;
	default:
		fprintf(stderr, "Failed to bind to NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out_free;
	}

	retval = nsdb_get_fsn_s(host, nce, fsn_uuid, &fsn, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		fsn_ttl = fsn->fn_fsnttl;
		nsdb_free_fedfs_fsn(fsn);
		break;
	case FEDFS_ERR_NSDB_NONCE:
		if (nce == NULL)
			fprintf(stderr, "NSDB %s:%u has no NCE\n",
				nsdbname, nsdbport);
		else
			fprintf(stderr, "NCE %s does not exist\n", nce);
		goto out_close;
	case FEDFS_ERR_NSDB_NOFSN:
		fprintf(stderr, "Failed to find FSN %s\n", fsn_uuid);
		goto out_close;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "NSDB LDAP error: %s\n",
			ldap_err2string(ldap_err));
		goto out_close;
	default:
		fprintf(stderr, "FedFsStatus code "
			"while retrieving FSN UUID %s: %s\n",
			fsn_uuid, nsdb_display_fedfsstatus(retval));
		goto out_close;
	}

	retval = nsdb_resolve_fsn_s(host, nce, fsn_uuid, &fsls, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		printf("For FSN UUID: %s\n", fsn_uuid);
		printf("    FSN TTL: %d\n\n", fsn_ttl);
		nsdb_resolve_fsn_display_fsls(fsls);
		nsdb_free_fedfs_fsls(fsls);
		break;
	case FEDFS_ERR_NSDB_NONCE:
		if (nce == NULL)
			fprintf(stderr, "NSDB %s:%u has no NCE\n",
				nsdbname, nsdbport);
		else
			fprintf(stderr, "NCE %s does not exist\n", nce);
		break;
	case FEDFS_ERR_NSDB_NOFSL:
		printf("For FSN UUID: %s\n", fsn_uuid);
		printf("    FSN TTL: %d\n", fsn_ttl);
		printf("    No FSL entries found\n");
		break;
	case FEDFS_ERR_NSDB_NOFSN:
		fprintf(stderr, "Failed to find FSN %s\n", fsn_uuid);
		break;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "NSDB LDAP error: %s\n",
			ldap_err2string(ldap_err));
		break;
	default:
		fprintf(stderr, "FedFsStatus code "
			"while resolving FSN UUID %s: %s\n",
			fsn_uuid, nsdb_display_fedfsstatus(retval));
	}

out_close:
	nsdb_close_nsdb(host);

out_free:
	nsdb_free_nsdb(host);

out:
	exit((int)retval);
}
